src/planet-a/SubdivTriangleMesh.h

```diff
@@ -0,0 +1,487 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "SubdivSkeleton.h"
+
+#include <osp/types.h>
+
+#include <memory>
+#include <variant>
+
+namespace planeta
+{
+
+// ID for all chunks, from 0 to m_chunkMax
+enum class ChunkId : uint16_t {};
+
+// ID for all shared vertices; from 0 to m_sharedMax
+enum class SharedVrtxId : uint32_t {};
+
+// Indices of vertices, unaware of vertex size;
+// from 0 to (m_chunkMax*m_chunkVrtxCount + m_sharedMax)
+enum class VertexId : uint32_t {};
+
+// IDs for any chunk's shared vertices; from 0 to m_chunkVrtxSharedCount
+enum class ChunkLocalSharedId : uint16_t {};
+
+// IDs for any chunk's fill vertices; from 0 to m_chunkVrtxSharedCount
+enum class ChunkLocalFillId : uint16_t {};
+
+using SharedVrtxStorage_t = osp::IdRefCount<SharedVrtxId>::Storage_t;
+
+//-----------------------------------------------------------------------------
+
+/**
+ * Convert XY coordinates to a triangular number index
+ *
+ *  0
+ *  1  2
+ *  3  4  5
+ *  6  7  8  9
+ * 10 11 12 13 14
+ * ...
+ * x = right; y = down; (0, 0) = 0
+ *
+ * @param x [in]
+ * @param y [in]
+ * @return triangular number
+ */
+constexpr int xy_to_triangular(int x, int y) noexcept
+{
+    return y * (y + 1) / 2 + x;
+};
+
+constexpr std::pair<ChunkLocalSharedId, bool> coord_to_shared(
+        uint16_t x, uint16_t y, uint16_t chunkEdgeVrtxCount) noexcept
+{
+    // Tests if (x,y) lies along right, bottom, or left edges of triangle
+    // branchless because why not
+    bool lft = (x == 0);
+    bool btm = !lft && (y == chunkEdgeVrtxCount);
+    bool rte = !btm && (x == y);
+
+    // Calculate chunk local shared ID. starts at 0 from the top of the chunk,
+    // and increases counterclockwise along the edge
+    uint32_t offset = lft * y
+                    + btm * (chunkEdgeVrtxCount + x)
+                    + rte * (chunkEdgeVrtxCount * 3 - x);
+
+    return {ChunkLocalSharedId(offset), rte || btm || lft};
+}
+
+//-----------------------------------------------------------------------------
+
+
+struct Chunk
+{
+    SkTriStorage_t m_skeletonTri;
+};
+
+class ChunkedTriangleMeshInfo
+{
+
+public:
+
+    // from wolfram alpha: "sequence 1 3 9 21 45 93"
+    // non-trivial to calculate. ~10 is over the practical limit for
+    // subdivision levels.
+    static inline constexpr std::array<uint16_t, 10> smc_minFansVsLevel =
+    {
+        1, 3, 9, 21, 45, 93, 189, 381, 765, 1533
```

Comment:
What's the algorithm that you used on wolfram alpha to get this sequence?

Even if you can't actually implement it as compile-time math with constexpr, having some rough C++ psuedocode would be helpful for comprehension.

Reply:
So these are the 'fan' triangles labeled in red, which increase by subdivision level:
![image](https://user-images.githubusercontent.com/14708882/140592274-8220beed-8f14-4979-ba5d-8a3becfc27a7.png)

Since I'm lazy, the first few numbers in the sequence can be plugged into wolfram alpha, then they do the black magic:
https://www.wolframalpha.com/input/?i=sequence+1+3+9+21+45+93+189

It's a recurrence relation like the fibonacci sequence. Honestly, I tried to calculate it compile time, which did work, but a single look-up table like this is hundreds of times more easy to understand. An imaginary C programmer laughed at me for wasting 3 hours.